File: app/src/main/java/io/madrona/madsqlite/Cursor.java
package io.madrona.madsqlite;

import java.io.Closeable;

@SuppressWarnings("WeakerAccess")
public final class Cursor implements Closeable {

    //region CONSTANTS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

    //region FIELDS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    private long nativePtr;

    //endregion

    //region INJECTED DEPENDENCIES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

    //region INJECTED VIEWS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

    //region CONSTRUCTOR ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    Cursor(long ptr) {
        nativePtr = ptr;
    }

    //endregion

    //region PRIVATE METHODS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

    //region PUBLIC METHODS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    public boolean moveToFirst() {
        return JniBridge.moveToFirst(nativePtr);
    }

    public boolean moveToPosition(int position) {
        return JniBridge.moveToPosition(nativePtr, position);
    }

    public int getCount() {
        return JniBridge.getCount(nativePtr);
    }

    public boolean moveToNext() {
        return JniBridge.moveToNext(nativePtr);
    }

    public boolean isAfterLast() {
        return JniBridge.isAfterLast(nativePtr);
    }

    public String getString(int columnIndex) {
        return JniBridge.getString(nativePtr, columnIndex);
    }

    public byte[] getBlob(int columnIndex) {
        return JniBridge.getBlob(nativePtr, columnIndex);
    }

    public long getInt(int columnIndex) {
        return JniBridge.getInt(nativePtr, columnIndex);
    }

    public double getReal(int columnIndex) {
        return JniBridge.getReal(nativePtr, columnIndex);
    }

    //endregion

    //region ACCESSORS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

    //region {Closeable} ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    @Override
    public void close() {
        JniBridge.closeCursor(nativePtr);
        nativePtr = 0L;
    }

    //endregion

    //region INNER CLASSES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    //endregion

}

File: app/src/main/cpp/Cursor.cpp
//
// Created by William Kamp on 10/10/16.
//

#include "Cursor.h"

//region Class Methods ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//endregion

//region Constructor ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Cursor::Cursor(sqlite3_stmt *statement) : statement(statement) {
    count = -1;
    count = evaluateCount();
}

Cursor::Cursor(Cursor &&curs) {
    statement = curs.statement;
    count = curs.count;
    position = curs.position;
    curs.statement = nullptr;
    curs.count = 0;
    curs.position = 0;
}

Cursor::~Cursor() {
    if (statement) {
        sqlite3_finalize(statement);
    }
}

//endregion

//region Public Methods ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

bool Cursor::moveToFirst() {
    if (sqlite3_reset(statement) == SQLITE_OK) {
        position = 0;
        return true;
    }
    return false;
}

int Cursor::getCount() {
    return count;
}

bool Cursor::moveToPosition(int p) {
    if (p < position) {
        moveToFirst();
    }
    while (position < p && moveToNext());
    return position == p;
}

bool Cursor::moveToNext() {
    if (!isAfterLast()) {
        if (sqlite3_step(statement) == SQLITE_ROW) {
            ++position;
            return true;
        }
    }
    return false;
}

bool Cursor::isAfterLast() {
    return position == count;
}

const std::string Cursor::getString(int columnIndex) const {
    const unsigned char* text = sqlite3_column_text(statement, columnIndex);
    if (text) {
        return std::string(reinterpret_cast<const char*>(text));
    }
    return "";
}

const std::vector<byte> Cursor::getBlob(int columnIndex) const {
    const void *blob = sqlite3_column_blob(statement, columnIndex);
    int sz = sqlite3_column_bytes(statement, columnIndex);
    const byte *charBuf = reinterpret_cast<const byte*>(blob);
    std::vector<byte> value(charBuf, charBuf + sz);
    return value;
}

uint64_t Cursor::getInt(int columnIndex) {
    return (uint64_t) sqlite3_column_int(statement, columnIndex);
}

double Cursor::getReal(int columnIndex) {
    return sqlite3_column_double(statement, columnIndex);
}

//endregion

//region Private Methods ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

int Cursor::evaluateCount() {
    while (moveToNext());
    int c = position;
    moveToFirst();
    return c;
}

//endregion
